### PR TITLE
Build conda on merge to main

### PR DESCRIPTION
**Pull request by ryan-s**
### Pull Request Description
Add in a conda package build as part of the ci when a PR is merged into main.  This PR adds in a small python script to change the recipe to use the local build (conda_config.py). Once we have that, we clone the feedstock and start to replicate what the python and shell scripts were doing do to issues on circleCI.  The end result is a build that passes or fails and produces no artifacts.

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

File: .circleci/config.yml
Comment (freddyaboulton) [inline .circleci/config.yml:60]: We have manually check out this container and run `docker run` ourselves as opposed to running `run_docker_build.sh` in the feedstock repo because circle-ci doesn't let you mount volumes to the docker image. I think this is ok because we're not straying far from the contents of that script. @ryan-s and I discussed ways of tightening this up in the future too!

File: .circleci/config.yml
Comment (freddyaboulton) [inline .circleci/config.yml:344]: This means that this will run after a PR gets merged to `main`.

We created this check to catch breaking changes to our conda recipe (like new dependencies or dependency updates) before release day and having it run after a merge accomplishes that without having to slow down development or consume too many ci resources.

**Review by freddyaboulton — APPROVED** [review]
@ryan-s this is awesome! I left some comments to explain to the team what's going on (feel free to correct me where I'm mistaken hehe).

We ran this check on circle ci with [main](https://app.circleci.com/pipelines/github/alteryx/evalml/6259/workflows/fe07b6f1-813d-4ccd-b70b-7ba1ffe3b996/jobs/67031) and a [purposely broken branch](https://app.circleci.com/pipelines/github/alteryx/evalml/6260/workflows/0b49c30d-a446-4329-8513-9bfe4b9de95e/jobs/67046) and the check is working as expected!

File: dev-requirements.txt
Comment (ryan-s) [reply dev-requirements.txt]: Moved

File: .circleci/config.yml
Comment (angela97lin) [reply .circleci/config.yml:344]: very cool!

Comment (dsherry) [reply .circleci/config.yml:344]: SGTM!

File: .circleci/config.yml
Comment (dsherry) [reply .circleci/config.yml:60]: @ryan-s @freddyaboulton is this env var `DOCKER_HUB_PASS` set by circleci? Did you have to enter it?

File: .circleci/config.yml
Comment (dsherry) [inline .circleci/config.yml:73]: If I'm following right, this command will have a non-zero return code if there's an error? Meaning the CI job will get flagged as failed?

Will the error output appear in the CI job, or do we have to do something else to get the error output? It would be amazing to have the output appear in the CI job, because otherwise its gonna be hard to debug.

File: .circleci/config.yml
Comment (dsherry) [inline .circleci/config.yml:85]: What is this version of? Is this the version of docker itself? And is `setup_remote_docker` a circleci predefined macro, or something else?

File: .circleci/config.yml
Comment (dsherry) [inline .circleci/config.yml:80]: @ryan-s @freddyaboulton why python 3.6? Should we do this for all our supported python versions, 3.6, 3.7 and 3.8? Or just the latest version, 3.8?

File: .circleci/config.yml
Comment (dsherry) [inline .circleci/config.yml:341]: Why is this listed as required when it just ran beforehand

File: conda_config.py
Comment (dsherry) [inline conda_config.py:38]: @ryan-s this is super cool!!

Please move this file to a different location instead of the top-level directory. Should it go in `.circleci/`? That would be my first inclination.

File: conda_config.py
Comment (dsherry) [inline conda_config.py:17]: Amazing, thank you for documenting.

File: conda_config.py
Comment (dsherry) [inline conda_config.py:7]: Why is this necessary? Is this just a style thing, or is it a functional thing?

File: test-requirements.txt
Comment (dsherry) [inline test-requirements.txt:10]: 👍

File: conda_config.py
Comment (dsherry) [inline conda_config.py:26]: @ryan-s could you please [use `pathlib.Path`](https://docs.python.org/3/library/pathlib.html) here instead? For hypothetical windows support haha

```python
recipe_path = str(pathlib.Path('..', 'feedstock_root', 'evalml'))
```

Also what is this path? Like, this is intended to be the path on the docker container? A comment explaining this would be super helpful to future readers.

**Review by dsherry — APPROVED** [review]
@ryan-s this rocks!!! I'm so excited you're adding this! It's gonna make our release process a lot simpler and less likely to break conda. Woohoo!!

So, I take it we'll still need to do the actual conda release manually, right?

@freddyaboulton once this is in, could you update release.md to reflect this?

@ryan-s I left a bunch of questions, and two comments to address before merge:
* Please move that new file `conda_config.py` to somewhere other than the top-level directory -- I think it should go in `.circleci/`. Hmm, on reflection, since its python code, perhaps it should go in a new folder `evalml/tests/circleci/`. Your call there.
* I left a comment about using `Pathlib`--let's resolve that, either by making the change or by you telling me its unnecessary lol

🚢

File: .circleci/config.yml
Comment (freddyaboulton) [reply .circleci/config.yml:73]: The error appears in the CI job!

File: .circleci/config.yml
Comment (freddyaboulton) [reply .circleci/config.yml:80]: I think this sets the version of python in the circleci worker. The docker run command will actually build the package in python 3.7.8 - this is the same as what happens in the CI of our feedstock repo. Unfortunately, there isn't an out-of-the-box way to build with different python versions but I will add this to the future work section of our design doc for this issue.

**Comment by freddyaboulton** [comment]
> @ryan-s this rocks!!! I'm so excited you're adding this! It's gonna make our release process a lot simpler and less likely to break conda. Woohoo!!
> 
> So, I take it we'll still need to do the actual conda release manually, right?
> 
> @freddyaboulton once this is in, could you update release.md to reflect this?
> 
> @ryan-s I left a bunch of questions, and two comments to address before merge:
> 
> * Please move that new file `conda_config.py` to somewhere other than the top-level directory -- I think it should go in `.circleci/`. Hmm, on reflection, since its python code, perhaps it should go in a new folder `evalml/tests/circleci/`. Your call there.
> * I left a comment about using `Pathlib`--let's resolve that, either by making the change or by you telling me its unnecessary lol
> 
> 🚢

Yes, we still need to manually edit the recipe file in the feedstock if there are dependency changes in a release but now we can skip the "Test conda version before releasing on PyPI" section of the release process! I'll update once this is in.

File: .circleci/config.yml
Comment (dsherry) [reply .circleci/config.yml:80]: Ah got it. So we'll only build and run our tests on 3.7.8 for now. No problem, this is a great starting point. Thanks!

File: .circleci/config.yml
Comment (ryan-s) [reply .circleci/config.yml:60]: @dsherry I entered it in circleci. Not sure if we have a vault somewhere to share the credentials?

File: .circleci/config.yml
Comment (ryan-s) [reply .circleci/config.yml:85]: Correct, this is the latest version of docker that circleci supports and it is predefined by them.

File: .circleci/config.yml
Comment (ryan-s) [reply .circleci/config.yml:80]: Correct, I just needed something to install pyyaml into so that we can update the recipe.  

I think the v2 of this would be to create our own container(s) based off of condaforge/linux-anvil-comp7 where we can move the existing functionality and any new improvements into.

File: conda_config.py
Comment (ryan-s) [reply conda_config.py:7]: Unfortunately it is functional.  The existing recipe is actually a Jinja2 template masquerading as a yaml file.  Whatever they are using to parse that template requires that we have the increased indents even though the YAML spec does not.

**Comment by ryan-s** [comment]
> @ryan-s this rocks!!! I'm so excited you're adding this! It's gonna make our release process a lot simpler and less likely to break conda. Woohoo!!
> 
> So, I take it we'll still need to do the actual conda release manually, right?
> 
> @freddyaboulton once this is in, could you update release.md to reflect this?
> 
> @ryan-s I left a bunch of questions, and two comments to address before merge:
> 
> * Please move that new file `conda_config.py` to somewhere other than the top-level directory -- I think it should go in `.circleci/`. Hmm, on reflection, since its python code, perhaps it should go in a new folder `evalml/tests/circleci/`. Your call there.
> * I left a comment about using `Pathlib`--let's resolve that, either by making the change or by you telling me its unnecessary lol
> 
> 🚢

As of right now yes, but in theory it wouldn't be too hard to copy the artifact out of the container and push it.  You would just need to do the inverse of what we do when we copy our feedstock into the container.

**Pull request merged**
